### PR TITLE
Added support of onCancellation handler for coroutine cancellation

### PR DESCRIPTION
**Pull request by alefedor**

### CODE REVIEW

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt [hunk 1]
```diff
@@ -200,11 +200,13 @@ fun <T> CancellableContinuation<T>.cancelByLincheck(promptCancellation: Boolean)
...
         // TODO we can invoke the method directly if the transformation is disabled
         getMethod(this, cancelCompletedResultMethod).invoke(this, null, cancellationByLincheckException)
-        return true
+        return CancellationResult.CANCELLED_COMPLETED_RESULT
     }
-    return cancelled
+    return if (cancelled) CancellationResult.CANCELLED else CancellationResult.FAILED
 }
 
+internal enum class CancellationResult { CANCELLED, CANCELLED_COMPLETED_RESULT, FAILED }
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt:208]: How about `CANCELLED_BEFORE_RESUMPTION, CANCELLED_AFTER_RESUMPTION, CANCELLATION_FAILED`?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt [hunk 2]
```diff
@@ -200,11 +200,13 @@ fun <T> CancellableContinuation<T>.cancelByLincheck(promptCancellation: Boolean)
     if (!cancelled && promptCancellation) {
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt:200]: Let's put `if (cancelled) return ...` here -- the code will be cleaner

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt [hunk 3]
```diff
@@ -225,14 +225,15 @@ internal open class ParallelThreadsRunner(
         return suspensionPointResults[iThread][actorId]
     }
 
+    internal open fun <T> cancelByLincheck(cont: CancellableContinuation<T>, promptCancellation: Boolean): CancellationResult =
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt:228]: Let's put a comment that we need to override it for trace construction

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt [hunk 4]
```diff
@@ -214,6 +216,23 @@ internal class UnparkTracePoint(
...
+    callStackTrace: CallStackTrace,
+) : TracePoint(iThread, actorId, callStackTrace) {
+    private lateinit var cancellationResult: CancellationResult
+
+    fun initializeCancellationResult(cancellationResult: CancellationResult) {
+        this.cancellationResult = cancellationResult
+    }
+
+    override fun toStringImpl(): String = when (cancellationResult) {
+        CancellationResult.CANCELLED -> "CANCELLED COROUTINE"
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt:230]: CANCELLED BEFORE RESUMPTION

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt [hunk 5]
```diff
@@ -214,6 +216,23 @@ internal class UnparkTracePoint(
...
+        CancellationResult.CANCELLED_COMPLETED_RESULT -> "CANCELLED COMPLETED COROUTINE"
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt:231]: PROMPT CANCELLED AFTER RESUMPTION

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt [hunk 6]
```diff
@@ -214,6 +216,23 @@ internal class UnparkTracePoint(
...
+        CancellationResult.FAILED -> "FAILED COROUTINE CANCELLATION"
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt:232]: CANCELLATION ATTEMPT FAILED

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyTransformer.kt [hunk 7]
```diff
@@ -950,23 +950,24 @@ internal class ManagedStrategyTransformer(
...
-     * when a continuation is cancelled by lincheck
+     * Removes ignored section for `CancellableContinuationImpl.onCancellation` call, so that
+     * events in the `onCancellation` method influence the execution and the trace.
+     *
+     * Note that this transformer does not handle the case of cancel handler throwing an exception.
      */
     private inner class CancellabilitySupportMethodTransformer(methodName: String, mv: GeneratorAdapter) : ManagedStrategyMethodVisitor(methodName, mv) {
-        private val isCancel = className == "kotlinx/coroutines/CancellableContinuationImpl" &&
-            (methodName == "cancel" || methodName == "cancelCompletedResult")
+        private val isOnCancellation = className == "kotlinx/coroutines/InvokeOnCancel" && methodName == "invoke"
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyTransformer.kt:959]: This is a private class -- we can't rely on it.

Comment (ndkoval) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyTransformer.kt:959]: We could leave and enter the ignored section back for the whole cancellation, I guess

**Review by ndkoval — CHANGES_REQUESTED** [review]

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyTransformer.kt (hunk 7 above)
Comment (alefedor) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyTransformer.kt]: I leaved ignored section here so as to log the least possible odd method calls.
I can do the same for the `CancellableContinutionImplcallCancelHandler`, but `onCancellation` call stack will look worse.

Now it is
```
invoke(Exception@1): Unit@1 at InvokeOnCancel.invoke(CancellableContinuationImpl.kt:539)
  invoke(Exception@1) at CoroutineCancellationTraceReportingTest$cancelledOp$$inlined$suspendCancellableCoroutine$lambda$1.invoke(CoroutineCancellationTraceReportingTest.kt:30)
    setCorrect(false) at CoroutineCancellationTraceReportingTest$cancelledOp$$inlined$suspendCancellableCoroutine$lambda$1.invoke(CoroutineCancellationTraceReportingTest.kt:39)
      correct.WRITE(false) at CoroutineCancellationTraceReportingTest.setCorrect(CoroutineCancellationTraceReportingTest.kt:32)
```
but will be
```
invoke(Exception@1) at CancellableContinuationImpl.callCancelHandler(CancellableContinuationImpl.kt:230)
  invoke(Exception@1): Unit@1 at InvokeOnCancel.invoke(CancellableContinuationImpl.kt:539)
    invoke(Exception@1) at CoroutineCancellationTraceReportingTest$cancelledOp$$inlined$suspendCancellableCoroutine$lambda$1.invoke(CoroutineCancellationTraceReportingTest.kt:30)
      setCorrect(false) at CoroutineCancellationTraceReportingTest$cancelledOp$$inlined$suspendCancellableCoroutine$lambda$1.invoke(CoroutineCancellationTraceReportingTest.kt:39)
        correct.WRITE(false) at CoroutineCancellationTraceReportingTest.setCorrect(CoroutineCancellationTraceReportingTest.kt:32)
```

By the way, it looks bad in both variants

Comment (alefedor) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyTransformer.kt]: Well, now it is
```
_state.compareAndSet(InvokeOnCancel@1,CancelledContinuation@1): true at CancellableContinuationImpl.cancel(CancellableContinuationImpl.kt:193)
invoke(Exception@1) at CancellableContinuationImpl.callCancelHandler(CancellableContinuationImpl.kt:230)
  invoke(Exception@1): Unit@1 at InvokeOnCancel.invoke(CancellableContinuationImpl.kt:539)
    invoke(Exception@1) at CoroutineCancellationTraceReportingTest$cancelledOp$$inlined$suspendCancellableCoroutine$lambda$1.invoke(CoroutineCancellationTraceReportingTest.kt:30)
      setCorrect(false) at CoroutineCancellationTraceReportingTest$cancelledOp$$inlined$suspendCancellableCoroutine$lambda$1.invoke(CoroutineCancellationTraceReportingTest.kt:39)
        correct.WRITE(false) at CoroutineCancellationTraceReportingTest.setCorrect(CoroutineCancellationTraceReportingTest.kt:32)

```

Comment (ndkoval) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyTransformer.kt]: Too much `invoke`-s  😄

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyTransformer.kt [hunk 8]
```diff
@@ -950,23 +950,24 @@ internal class ManagedStrategyTransformer(
...
 
     /**
-     * Removes switch points in CancellableContinuationImpl.cancel, so that they will not be reported
-     * when a continuation is cancelled by lincheck
+     * Removes ignored section for `CancellableContinuationImpl.onCancellation` call, so that
+     * events in the `onCancellation` method influence the execution and the trace.
+     *
+     * Note that this transformer does not handle the case of cancel handler throwing an exception.
      */
     private inner class CancellabilitySupportMethodTransformer(methodName: String, mv: GeneratorAdapter) : ManagedStrategyMethodVisitor(methodName, mv) {
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyTransformer.kt:958]: Can we remove this transformer and add `invokeAfterIgnoredSectionLeaving` and `invokeBeforeIgnoredSectionEntering` invocations to the runner's `cancelByLincheck` implementation?

**Review by ndkoval — CHANGES_REQUESTED** [review]

**Review by ndkoval — APPROVED** [review]
Super!

**Pull request merged**